File: backend/internal/modules/recrutamento/handlers/tenant_signature.go
package handlers

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"net/http"
	"strconv"
	"time"
)

// tenantSignatureMaxSkew é a janela de validade de X-Tenant-Timestamp —
// impede reutilização (replay) de uma assinatura capturada.
const tenantSignatureMaxSkew = 5 * time.Minute

var errInvalidTenantSignature = errors.New("assinatura de tenant inválida")

// resolveTenantBySignature identifica o tenant através de um HMAC-SHA256
// assinado com saas.tenants.recrutamento_api_secret, para chamadores que não
// podem confiar no Host/X-Forwarded-Host mediado pelo Traefik — ex.: um site
// institucional fora da infra do Nexora que chama a API pública directamente
// do lado do servidor. Ao contrário de X-Forwarded-Host (ver requestHost),
// este cabeçalho não é forjável sem conhecer o segredo do tenant.
//
// Cabeçalhos esperados:
//
//	X-Tenant-Code:      código do tenant (saas.tenants.codigo)
//	X-Tenant-Timestamp: unix seconds do pedido
//	X-Tenant-Signature:  hex(HMAC-SHA256(segredo, "<codigo>.<timestamp>.<method>.<path>"))
func (h *Handler) resolveTenantBySignature(r *http.Request, codigo string) (int64, error) {
	tsRaw := r.Header.Get("X-Tenant-Timestamp")
	sigRaw := r.Header.Get("X-Tenant-Signature")
	if tsRaw == "" || sigRaw == "" {
		return 0, errInvalidTenantSignature
	}

	ts, err := strconv.ParseInt(tsRaw, 10, 64)
	if err != nil {
		return 0, errInvalidTenantSignature
	}
	if skew := time.Since(time.Unix(ts, 0)); skew > tenantSignatureMaxSkew || skew < -tenantSignatureMaxSkew {
		return 0, errInvalidTenantSignature
	}

	sig, err := hex.DecodeString(sigRaw)
	if err != nil {
		return 0, errInvalidTenantSignature
	}

	var id int64
	var secret *string
	err = h.db.QueryRow(r.Context(),
		`SELECT id, recrutamento_api_secret FROM saas.tenants WHERE codigo=$1 AND status='ativo'`,
		codigo,
	).Scan(&id, &secret)
	if err != nil || secret == nil || *secret == "" {
		return 0, errInvalidTenantSignature
	}

	mensagem := codigo + "." + tsRaw + "." + r.Method + "." + r.URL.Path
	mac := hmac.New(sha256.New, []byte(*secret))
	mac.Write([]byte(mensagem))

	if !hmac.Equal(sig, mac.Sum(nil)) {
		return 0, errInvalidTenantSignature
	}

	return id, nil
}

File: frontend/src/Infrastructure/Nexora/TenantSignature.php
<?php

declare(strict_types=1);

namespace E258Tech\Infrastructure\Nexora;

/**
 * Gera cabeçalhos de assinatura HMAC-SHA256 para identificação de tenant
 * nos endpoints públicos do Nexora ERP.
 *
 * Contrato (espelha o backend em internal/modules/recrutamento/handlers/tenant_signature.go):
 *
 *   X-Tenant-Code:      código do tenant (saas.tenants.codigo)
 *   X-Tenant-Timestamp: unix seconds do pedido
 *   X-Tenant-Signature: hex(HMAC-SHA256(segredo, "<codigo>.<timestamp>.<method>.<path>"))
 *
 * Se NENHUM segredo estiver configurado, as chamadas públicas continuam a usar
 * X-Forwarded-Host (mecanismo de transição/desenvolvimento).
 */
final class TenantSignature
{
    public function __construct(
        private readonly ?string $tenantCode,
        private readonly ?string $tenantSecret
    ) {
    }

    /**
     * @return array<int, string> cabeçalhos HTTP a adicionar ao pedido
     */
    public function headers(string $method, string $path): array
    {
        if ($this->tenantCode === null || $this->tenantCode === ''
            || $this->tenantSecret === null || $this->tenantSecret === '') {
            return [];
        }

        $timestamp = (string) time();
        $message = $this->tenantCode . '.' . $timestamp . '.' . strtoupper($method) . '.' . $path;
        $signature = hash_hmac('sha256', $message, $this->tenantSecret);

        return [
            'X-Tenant-Code: ' . $this->tenantCode,
            'X-Tenant-Timestamp: ' . $timestamp,
            'X-Tenant-Signature: ' . $signature,
        ];
    }

    public function isConfigured(): bool
    {
        return ($this->tenantCode ?? '') !== '' && ($this->tenantSecret ?? '') !== '';
    }
}
